src/csvlook.c

```diff
@@ -0,0 +1,260 @@
+        config.source = fopen(argv[optind], "r");
+        if (!config.source) {
+            fprintf(stderr, "Could not open file %s for reading\n", argv[optind]);
+            exit(1);
+        }
+    }
+
+    LOG_D("%s\n","Done parsing config params");
+
+    _tokenizer = setup_tokenizer(config.separator, _buffer, _cells, CELL_BUFFER_SIZE);
+}
+
+static void finish_config(size_t cells_found) {
+
+    Cell const* current_cell = _cells;
+
+    while (current_cell < (_cells + cells_found) && current_cell->start != NULL) {
+        current_cell++;
+    }
+    config.column_count = (int)(current_cell - _cells);
+    config.column_sizes = calloc(sizeof(int), config.column_count);
+    config.final_column_sizes = calloc(sizeof(int), config.column_count);
+
+    const char* new_line = _cells[config.column_count-1].start + _cells[config.column_count - 1].length;
+    config.newline[0] = new_line[0];
+    config.newline_length = 1;
+    if (new_line[1] == '\n' || new_line[0] == '\r') {
+        config.newline[1] = '\n';
+        config.newline_length = 2;
+    }
+
+    LOG_V("%s\n", "Starting column width detection");
+    current_cell = _cells;
+    Cell* cell_end = _cells + cells_found;
+    int column_idx = 0;
+    while (current_cell < cell_end) {
+        if (current_cell->start == NULL) {
+            column_idx = 0;
+            LOG_V("%s\n","New Row");
+        }
+        else {
+            LOG_V("Processing column %d: Prev max %d Cur len %d\n", column_idx, config.column_sizes[column_idx], (int)current_cell->length);
+            /* max(prevmax, curlength) */
+            config.column_sizes[column_idx] = (config.column_sizes[column_idx] < (int)current_cell->length) ? (int)current_cell->length : config.column_sizes[column_idx];
+            /* min(curmax, configmax) */
+            config.column_sizes[column_idx] = (config.column_sizes[column_idx] > preconfig.max_width) ? preconfig.max_width : config.column_sizes[column_idx];
+            column_idx++;
+        }
+        current_cell++;
+    }
+    column_idx = 0;
+    LOG_D("%s\n", "Trying to parse specified widths");
+    char *token = strtok(preconfig.specified_widths, ",");
+    while (token != NULL) {
+        LOG_D("Got token %s\n", token);
+        if (strlen(token) > 0) {
+            int len = atoi(token);
+            if (len > 0) {
+                config.column_sizes[column_idx] = len;
+            }
+        }
+
+        column_idx++;
+        token = strtok(NULL, ",");
+    }
+
+    LOG_D("%s\n","Done finishing config calcs");
+    for (int i = 0; i < config.column_count; i++) {
+        LOG_D("Column %d: %d\n", i, config.column_sizes[i]);
+    }
+}
+
+static void output_cells(size_t cells_found, bool last_full) {
+    Cell* current_cell = _cells;
+    Cell* cell_end = _cells + cells_found;
+    int column_idx = 0;
+
+    while (current_cell < cell_end) {
+        if (current_cell->start == NULL) {
+            fprintf(stdout, "%.*s", (int)config.newline_length, config.newline);
```

Comment:
this could also use `fwrite(config.newline, sizeof(char), config.newline_length, stdout);`